frontend/src/host_orchestrator/orchestrator/instancemanager.go

```diff
@@ -220,55 +224,74 @@ func (m *CVDToolInstanceManager) launchCVD(req apiv1.CreateCVDRequest, op apiv1.
 
 const (
 	// TODO(b/267525748): Make these values configurable.
-	defaultBranch = "aosp-master"
-	defaultTarget = "aosp_cf_x86_64_phone-userdebug"
+	mainBuildDefaultBranch = "aosp-master"
+	mainBuildDefaultTarget = "aosp_cf_x86_64_phone-userdebug"
 )
 
 func (m *CVDToolInstanceManager) launchFromAndroidCI(
 	req apiv1.CreateCVDRequest, op apiv1.Operation) (*apiv1.CVD, error) {
-	var build apiv1.AndroidCIBuild
-	if req.CVD.BuildSource != nil &&
-		req.CVD.BuildSource.AndroidCIBuildSource != nil &&
-		req.CVD.BuildSource.AndroidCIBuildSource.MainBuild != nil {
-		build = *req.CVD.BuildSource.AndroidCIBuildSource.MainBuild
-	}
-	branch := build.Branch
-	if branch == "" {
-		branch = defaultBranch
+	var mainBuild, kernelBuild *apiv1.AndroidCIBuild = &apiv1.AndroidCIBuild{}, nil
+	if req.CVD.BuildSource != nil && req.CVD.BuildSource.AndroidCIBuildSource != nil {
+		buildSource := req.CVD.BuildSource.AndroidCIBuildSource
+		if buildSource.MainBuild != nil {
+			mainBuild = &(*req.CVD.BuildSource.AndroidCIBuildSource.MainBuild)
+		}
+		kernelBuild = buildSource.KernelBuild
 	}
-	target := build.Target
-	if target == "" {
-		target = defaultTarget
+	if err := updateBuildWithDefaults(mainBuild, m.buildAPI, mainBuildDefaultBranch, mainBuildDefaultTarget); err != nil {
+		return nil, err
 	}
-	buildID := build.BuildID
-	if buildID == "" {
-		var err error
-		buildID, err = m.buildAPI.GetLatestGreenBuildID(branch, target)
-		if err != nil {
-			return nil, err
+	var mainBuildDir, kernelBuildDir string
+	var mainBuildDwErr, kernelBuildDwErr error
+	downloads := []struct {
+		resultDir  *string
+		resultErr  *error
+		build      *apiv1.AndroidCIBuild
+		bundleType artifactsBundleType
+	}{
+		{&mainBuildDir, &mainBuildDwErr, mainBuild, cvdBundleType},
+		{&kernelBuildDir, &kernelBuildDwErr, kernelBuild, kernelBundleType},
+	}
+	var wg sync.WaitGroup
+	for _, d := range downloads {
+		if d.build != nil {
+			wg.Add(1)
+			d := d
+			go func() {
+				defer wg.Done()
+				dir, err := m.artifactsMngr.Download(d.build.BuildID, d.build.Target, d.bundleType)
```

Comment:
Is it mandatory to provide a kernel build id to launch from Android CI? I don't see where we enforce it or allow for one not to be downloaded.

Reply:
it's not mandatory, the code gets executed only if the relevant build reference is not nil